pkg/ruler/manager.go

```diff
@@ -0,0 +1,231 @@
+package ruler
+
+import (
+	"context"
+	"net/http"
+	"sync"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	ot "github.com/opentracing/opentracing-go"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/notifier"
+	promRules "github.com/prometheus/prometheus/rules"
+	"github.com/weaveworks/common/user"
+	"golang.org/x/net/context/ctxhttp"
+
+	store "github.com/cortexproject/cortex/pkg/ruler/rules"
+	"github.com/cortexproject/cortex/pkg/util"
+)
+
+type MultiTenantManager interface {
+	SyncRuleGroups(ctx context.Context, ruleGroups map[string]store.RuleGroupList)
+	GetRules(userID string) []*promRules.Group
+	Stop()
+}
+
+type DefaultMultiTenantManager struct {
+	cfg            Config
+	notifierCfg    *config.Config
+	managerFactory ManagerFactory
+
+	mapper *mapper
+
+	// Structs for holding per-user Prometheus rules Managers
+	// and a corresponding metrics struct
+	userManagerMtx     sync.Mutex
+	userManagers       map[string]*promRules.Manager
+	userManagerMetrics *ManagerMetrics
+
+	// Per-user notifiers with separate queues.
+	notifiersMtx sync.Mutex
+	notifiers    map[string]*rulerNotifier
+
+	managersTotal prometheus.Gauge
+	registry      prometheus.Registerer
+	logger        log.Logger
+}
+
+func NewDefaultMultiTenantManager(cfg Config, managerFactory ManagerFactory, reg prometheus.Registerer, logger log.Logger) (*DefaultMultiTenantManager, error) {
+	ncfg, err := buildNotifierConfig(&cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	userManagerMetrics := NewManagerMetrics()
+	if reg != nil {
+		reg.MustRegister(userManagerMetrics)
+	}
+
+	managersTotal := prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: "cortex",
+		Name:      "ruler_managers_total",
+		Help:      "Total number of managers registered and running in the ruler",
+	})
+	reg.MustRegister(managersTotal)
```

Comment:
You can use `promauto.With(reg).NewGauge(...)` which handles the case when `reg` is `nil` (checked few lines earlier).